src/main/scala/org/ergoplatform/nodeView/mempool/ErgoMemPool.scala

```diff
@@ -1,57 +1,115 @@
+    new ErgoMemPool(newPool, invalidated, settings)
   }
 
   override def remove(tx: ErgoTransaction): ErgoMemPool = {
-    unconfirmed.remove(tx.id)
-    this
+    new ErgoMemPool(unconfirmed - weighted(tx), invalidated, settings)
   }
 
-  override def filter(condition: (ErgoTransaction) => Boolean): ErgoMemPool = {
-    unconfirmed.retain { case (k, v) =>
+  override def filter(condition: ErgoTransaction => Boolean): ErgoMemPool = {
+    val newPool = unconfirmed.filter { case (_, v) =>
       condition(v)
     }
-    this
+    new ErgoMemPool(newPool, invalidated, settings)
+  }
+
+  def invalidate(tx: ErgoTransaction): ErgoMemPool = {
+    new ErgoMemPool(unconfirmed - weighted(tx), updateInvalidatedWith(tx), settings)
+  }
+
+  def putIfValid(tx: ErgoTransaction, state: ErgoState[_]): (ErgoMemPool, ProcessingOutcome) = {
+    state match {
+      case validator: TransactionValidation[ErgoTransaction@unchecked]
+        if !invalidated.contains(tx.id) && !unconfirmed.contains(weighted(tx)) &&
+          (unconfirmed.size < settings.nodeSettings.mempoolCapacity ||
+          weighted(tx)._2 > unconfirmed.firstKey._2) =>
+        validator.validate(tx).fold(
+          new ErgoMemPool(unconfirmed, updateInvalidatedWith(tx), settings) -> ProcessingOutcome.Invalidated(_),
+          _ => new ErgoMemPool(updatePoolWith(tx, unconfirmed), invalidated, settings) -> ProcessingOutcome.Accepted
+        )
+      case _ =>
+        this -> ProcessingOutcome.Declined
+    }
+  }
+
+  private def weighted(tx: ErgoTransaction): (ModifierId, Long) = {
+    val fee = tx.outputs
+      .filter(b => java.util.Arrays.equals(b.propositionBytes, propositionBytes))
+      .map(_.value)
+      .sum
+    tx.id -> fee
   }
-}
 
+  /**
+    * Updated memory pool with new transaction, if pool ran out of size limit then
+    * transaction with the smallest weight is replaced by the new one
+    */
+  private def updatePoolWith(tx: ErgoTransaction,
+                             pool: TreeMap[(ModifierId, Long), ErgoTransaction]) = {
+    (if (pool.size >= settings.nodeSettings.mempoolCapacity) pool - pool.firstKey else pool)
+      .updated(weighted(tx), tx)
+  }
+
+  private def updateInvalidatedWith(tx: ErgoTransaction) = {
+    (if (invalidated.size >= blacklistCapacity) invalidated - invalidated.toList.minBy(_._2)._1 else invalidated)
+      .updated(tx.id, System.currentTimeMillis)
+  }
+
+}
 
 object ErgoMemPool {
 
+  sealed trait ProcessingOutcome
+
+  object ProcessingOutcome {
+    case object Accepted extends ProcessingOutcome
+    case object Declined extends ProcessingOutcome
+    case class Invalidated(e: Throwable) extends ProcessingOutcome
+  }
+
   type MemPoolRequest = Seq[ModifierId]
 
   type MemPoolResponse = Seq[ErgoTransaction]
 
-  def empty: ErgoMemPool = new ErgoMemPool(TrieMap.empty)
+  private implicit val ord: Ordering[(ModifierId, Long)] = Ordering[(Long, ModifierId)].on(x => (x._2, x._1))
```

Comment:
@oskin1  I agree with @mike-aksarin, we probably need a class for this key. BTW, which the order on the right is reversed?

Reply:
@ergomorphic, we need keys to be sorted by weight first and then by ModifierId